### PR TITLE
chore(examples): Add navigation examples

### PR DESCRIPTION
**Pull request by gusvargas**
Related to https://github.com/redux-observable/redux-observable/issues/87

First stab at some navigation examples. It's a small app that let's you search for GH users. Clicking on a GH user gives you links to their repos.

The epic examples are:

1. Side loading data by responding to navigation actions
2. Canceling the side loading of data due to an action originating from a navigation
3. Restricting routes/delayed redirects

I'm relatively new to Rx so I'd appreciate any input on how to write more idiomatic use of operators. Also happy to work in any other ideas for navigation-related epics.

### CODE REVIEW

**Comment by jayphelps** [comment]
This is rad. Will dig into it when I have a moment 👍

File: examples/navigation/containers/Admin.js [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import React from 'react';
+
+export default function Admin() {
+  return (
+    <p>
+      You don't have access. Redirecting back home.
+    </p>
+  );
+}
```
Comment (blesh) [inline examples/navigation/containers/Admin.js]: You'll want to rename these files to `.jsx`, if for no other reason than to get GitHub support for JSX formatting.

File: examples/navigation/epics/adminAccess.js [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+function hasAccess() {
+  return false;
+}
+
+export default function adminAccess(action$) {
+  return action$.ofType(LOCATION_CHANGE)
+    .map(action => action.payload.pathname)
+    .filter(pathname => pathname === '/admin')
+    .map(hasAccess)
+    .filter(access => !access)
```
Comment (blesh) [inline examples/navigation/epics/adminAccess.js]: I don't understand the purpose of these two lines:

```js
.map(hasAccess) // hasAccess always returns false.
.filter(access => !access) // ...so this will allow everything through
```

They seem like they can be omitted.

File: examples/navigation/epics/clearSearchResults.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+import * as ActionTypes from '../actionTypes';
+
+export default action$ =>
+  action$.ofType(LOCATION_CHANGE)
+    .filter(({ payload }) =>
+      payload.pathname === '/' && !!!payload.query.q
+    )
+    .mapTo({
+      type: ActionTypes.CLEARED_RESULTS
+    });
```
Comment (blesh) [inline examples/navigation/epics/clearSearchResults.js]: Something to note here about `mapTo`, it will literally emit the same instance of the object over and over. If anything downstream is checking to see if the action is a new instance, it could have consequences. In contrast, `map(() => { foo: 'bar' })` will emit a new object instance each time.

File: examples/navigation/epics/fetchReposByUser.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    fetchIntents$
+      .switchMap(user =>
+        ajax.getJSON(`https://api.github.com/users/${user}/repos`)
+          .map(res => ({
+            type: ActionTypes.RECEIVED_USER_REPOS,
+            payload: {
+              repos: res,
+              user
+            }
+          }))
```
Comment (blesh) [inline examples/navigation/epics/fetchReposByUser.js]: It's slightly more performant to use the mapping function that is the second argument to `switchMap`:

```js
switchMap((user) => ajax.getJSON(`https://api.github.com/users/${user}/repos`),
                   (user, res) => ({
                     type: ActionTypes.RECEIVED_USER_REPOS,
                     payload: {
                       repos: res,
                       user
                     }
                   }))
```

File: examples/navigation/index.js [hunk 5]
```diff
@@ -0,0 +1,30 @@
+import 'rxjs';
```
Comment (blesh) [inline examples/navigation/index.js]: This is interesting. Why are we just script-importing `rxjs` here?

File: examples/navigation/package.json [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+  "main": "index.js",
+  "scripts": {
+    "serve": "webpack-dev-server --inline --history-api-fallback",
+    "build": "webpack"
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/redux-observable/redux-observable.git"
+  },
+  "author": "",
```
Comment (blesh) [inline examples/navigation/package.json]: BRO. Your name here.

File: examples/navigation/epics/adminAccess.js (hunk 2 above)
Comment (gusvargas) [inline examples/navigation/epics/adminAccess.js]: Sorry, I should have left a comment. Those dummy lines are to imply that more meaningful code should go there to *actually* check for access – it's just for semantics. I could leave a comment to the existing code or just omit it all together.

File: examples/navigation/epics/clearSearchResults.js (hunk 3 above)
Comment (gusvargas) [inline examples/navigation/epics/clearSearchResults.js]: Ah – I never considered that that anyone would care about the identity of actions. Good point.

File: examples/navigation/epics/fetchReposByUser.js (hunk 4 above)
Comment (jayphelps) [inline examples/navigation/epics/fetchReposByUser.js]: @blesh I don't believe we want to advocate that for these sort of circumstances because it will confuse them and create bugs when they want to add error handling. e.g. if they put a `.catch()` that returns an error action inside the switchMap, now will wrongfully be mapped as the success response. If they place that `.catch()` outside the switchMap, it is too late and will prevent their Epic from listening to future actions.

File: examples/navigation/index.js (hunk 5 above)
Comment (jayphelps) [inline examples/navigation/index.js]: @blesh it is done so that all the operators are placed on `Observable.prototype` because Epics receive an existing observable that they do not create. This is suggested in the docs: http://redux-observable.js.org/docs/basics/Epics.html#a-basic-example

Comment (gusvargas) [inline examples/navigation/index.js]: This is originally why I added it – the operators I wanted were not on the `ActionsObservable` and I was immediately sure which operators I was going to want to use. However, I just removed it and now *Just Works*.

Comment (jayphelps) [inline examples/navigation/index.js]: @gusvargas it shouldn't work.... o.O

Comment (jayphelps) [inline examples/navigation/index.js]: @gusvargas OH it works because you're effectively doing the same thing in `searchUsers.js`

https://github.com/redux-observable/redux-observable/pull/89/files/82ef94d41941f5d35b6bdd7a834bd13ffdc7165f#diff-98f4ddf4496646ae41dc030ac31bb6b5R1

You're importing the root `rxjs` to get `Observable`..the root `rx` imports _every_ operator and puts them on the prototype.

I would keep the `import 'rxjs';` in the index.js because it's the safest way to say "the app gets all of RxJS"

File: examples/navigation/epics/adminAccess.js (hunk 2 above)
Comment (blesh) [inline examples/navigation/epics/adminAccess.js]: I would omit the lines of code, but add a comment telling the reader what they _could_ do there.

File: examples/navigation/index.js (hunk 5 above)
Comment (gusvargas) [inline examples/navigation/index.js]: Ah – I see. Thanks for clearing that up.

File: examples/navigation/epics/fetchReposByUser.js (hunk 4 above)
Comment (blesh) [inline examples/navigation/epics/fetchReposByUser.js]: >  if they put a .catch() that returns an error action inside the switchMap, now will wrongfully be mapped as the success response

I can only assume you mean they're trying to handle some error that could have occurred during the mapping of the response? If so, then you're right.

I'm mostly indifferent, but *I personally* would like people to know about the projection function on `switchMap`, `flatMap`, `concatMap`, etc.

Then again, when/if I ever get the sync operations flattening to land, it shouldn't matter as much. There will still be an added closure, which will slow things down, but it will work.

I retract my suggestion. Mostly because Jay cares more than I do. haha.

File: examples/navigation/index.js (hunk 5 above)
Comment (blesh) [inline examples/navigation/index.js]: I see. That makes sense. I'd add a comment though, because it reads funny to those of us who are used to only importing exactly what we need from rxjs.

**Comment by gusvargas** [comment]
@jayphelps @blesh I've addressed all the comments so far. Any other suggestions?

File: examples/navigation/index.js (hunk 5 above)
Comment (gusvargas) [inline examples/navigation/index.js]: Comment added in https://github.com/redux-observable/redux-observable/pull/89/commits/e885d0ad45f73b9c9405a9044a029a179a3f11cb

**Comment by blesh** [comment]
I think this example is sizing up to be great. I feel like I've learned things.

**Comment by blesh** [comment]
LGTM :shipit:

**Comment by gusvargas** [comment]
Had some side conversation with @jayphelps about a couple things, mainly decoupling data fetching from the router action and having the intent come form the container component. I'll take make those changes soon.

**Comment by blesh** [comment]
> Had some side conversation with @jayphelps about a couple things

Oh... it would be good for you two to try to document/summarize these sorts of discussions on the issues next time (no need this time).

**Comment by jayphelps** [comment]
@blesh agreed. I usually do this, just didn't because the discussion was lengthy and forgot to give a short summary.

tl;dr for numerous lengthy reasons we agreed that the epics should listen for a component dispatched actions instead of `LOCATION_CHANGE`.

**Comment by blesh** [comment]
> tl;dr for numerous lengthy reasons we agreed that the epics should listen for a component dispatched actions instead of LOCATION_CHANGE.

That's a bummer, because that was the interesting thing I learned from this example. :\

**Comment by jayphelps** [comment]
@blesh maybe you'll convince me otherwise 😄

**Comment by gusvargas** [comment]
These are the highlights of that discussion in case anyone else has something to add:

- Coupling the URL to the epic adds yet another place that would need to be updated if paths were ever to change.
- In the case of `react-router-redux` it doesn't hook into `react-router` so you have to roll your own code for extracting path params.
- You can't be consistent with the pattern because not every container component's data needs are parameterized in the URL.
- However, it is really nice that there is this implicit machinery that understands when data should be fetched.

One alternative would be for container components to explicitly request their data needs by firing off an action. This approach comes with a boilerplate issue, i.e., both `componentDidMount` and `componentWillReceiveProps` would need to be implemented to capture all cases where data may need to be fetched. This is boilerplate that could be abstracted away but that's outside the scope of this PR.

Neither of us are certain that one pattern is better than the other but we're leaning towards the latter since it seems like even without an abstraction to reduce the boilerplate it will still result in less brittle/inconsistent code than the former.

**Comment by gusvargas** [comment]
@jayphelps @blesh I updated this PR per the points listed above.

**Comment by jayphelps** [comment]
Started reviewing, will finish soon. So far it looks super sexy. 😎

**Comment by jayphelps** [comment]
🎉 Huge thank you @gusvargas !!

**Comment by blesh** [comment]
Awesomeness, @gusvargas! 🍻
